Predefined documents do not load on Windows

When starting the CMV server, the server attempts to discover predefined documents from the classpath. On Linux and macOS this works fine, but on Windows the following error is reported:

```text
2024-06-26 13:51:24.875  WARN 30156 --- [           main] eu.cessda.cmv.server.Server              : Couldn't discover demo documents

java.lang.IllegalArgumentException: java.net.URISyntaxException: Illegal character in opaque part at index 12: jar:file:/C:\Users\MatthewMorris\.m2\repository\eu\cessda\cmv\cmv-core\3.0.0\cmv-core-3.0.0.jar!/demo-documents/ddi-v25/ukds-7481.xml
	at org.gesis.commons.resource.Resource.newResource(Resource.java:150)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:197)
	at java.base/java.util.stream.ReferencePipeline$2$1.accept(ReferencePipeline.java:179)
	at java.base/java.util.ArrayList$ArrayListSpliterator.forEachRemaining(ArrayList.java:1708)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:509)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:499)
	at java.base/java.util.stream.ReduceOps$ReduceOp.evaluateSequential(ReduceOps.java:921)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.collect(ReferencePipeline.java:682)
	at org.gesis.commons.resource.ClasspathResourceRepository.<init>(ClasspathResourceRepository.java:40)
	at org.gesis.commons.resource.ClasspathResourceRepository.<init>(ClasspathResourceRepository.java:11)
	at org.gesis.commons.resource.ClasspathResourceRepository$Builder.build(ClasspathResourceRepository.java:61)
	at eu.cessda.cmv.server.Server.demoDocuments(Server.java:132)
	at eu.cessda.cmv.server.Server$$EnhancerBySpringCGLIB$$3217a54c.CGLIB$demoDocuments$4(<generated>)
	at eu.cessda.cmv.server.Server$$EnhancerBySpringCGLIB$$3217a54c$$FastClassBySpringCGLIB$$67751ea6.invoke(<generated>)
	at org.springframework.cglib.proxy.MethodProxy.invokeSuper(MethodProxy.java:244)
	at org.springframework.context.annotation.ConfigurationClassEnhancer$BeanMethodInterceptor.intercept(ConfigurationClassEnhancer.java:331)
	at eu.cessda.cmv.server.Server$$EnhancerBySpringCGLIB$$3217a54c.demoDocuments(<generated>)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:648)
	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:481)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1352)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1195)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:582)
	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:542)
	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:335)
	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:234)
	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:333)
	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:208)
	at org.springframework.beans.factory.support.DefaultListableBeanFactory.preInstantiateSingletons(DefaultListableBeanFactory.java:955)
	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:929)
	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:591)
	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:147)
	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:732)
	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:409)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:308)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1300)
	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1289)
	at eu.cessda.cmv.server.Server.main(Server.java:59)
Caused by: java.net.URISyntaxException: Illegal character in opaque part at index 12: jar:file:/C:\Users\MatthewMorris\.m2\repository\eu\cessda\cmv\cmv-core\3.0.0\cmv-core-3.0.0.jar!/demo-documents/ddi-v25/ukds-7481.xml
	at java.base/java.net.URI$Parser.fail(URI.java:2995)
	at java.base/java.net.URI$Parser.checkChars(URI.java:3166)
	at java.base/java.net.URI$Parser.parse(URI.java:3202)
	at java.base/java.net.URI.<init>(URI.java:645)
	at java.base/java.net.URL.toURI(URL.java:1220)
	at org.gesis.commons.resource.Resource.newResource(Resource.java:146)
	... 40 common frames omitted
```

This is due to a bug in `org.gesis.commons.resource.Resource.newResource()` when trying to convert the `URL` into a `URI`.

--- a/src/main/java/eu/cessda/cmv/server/Server.java
+++ b/src/main/java/eu/cessda/cmv/server/Server.java
@@ -28,24 +28,30 @@ import com.fasterxml.jackson.module.jaxb.JaxbAnnotationModule;
 import eu.cessda.cmv.core.CessdaMetadataValidatorFactory;
 import eu.cessda.cmv.core.NotDocumentException;
 import eu.cessda.cmv.core.Profile;
-import org.gesis.commons.resource.ClasspathResourceRepository;
 import org.gesis.commons.resource.Resource;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.boot.SpringApplication;
 import org.springframework.boot.autoconfigure.SpringBootApplication;
 import org.springframework.boot.builder.SpringApplicationBuilder;
 import org.springframework.boot.web.servlet.support.SpringBootServletInitializer;
+import org.springframework.context.ApplicationContext;
 import org.springframework.context.annotation.Bean;
+import org.springframework.core.io.UrlResource;
+import org.springframework.util.StringUtils;
 import org.springframework.web.cors.CorsConfiguration;
 import org.springframework.web.cors.UrlBasedCorsConfigurationSource;
 import org.springframework.web.filter.CorsFilter;
 import org.zalando.problem.jackson.ProblemModule;
 
 import java.io.IOException;
-import java.util.Collections;
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashSet;
 import java.util.List;
-import java.util.stream.Stream;
 
 import static org.gesis.commons.resource.Resource.newResource;
 
@@ -53,6 +59,14 @@ import static org.gesis.commons.resource.Resource.newResource;
 public class Server extends SpringBootServletInitializer
 {
 	private static final Logger log = LoggerFactory.getLogger( Server.class );
+
+	private final ApplicationContext applicationContext;
+
+	@Autowired
+	public Server( ApplicationContext applicationContext )
+	{
+		this.applicationContext = applicationContext;
+	}
 
 	public static void main( String[] args )
 	{
@@ -83,28 +97,27 @@ public class Server extends SpringBootServletInitializer
 	}
 
 	@Bean
-	public List<Profile> demoProfiles( CessdaMetadataValidatorFactory factory )
+	public List<Profile> demoProfiles( CessdaMetadataValidatorFactory factory ) throws IOException
 	{
 		log.info( "Loading built-in profiles" );
-		return ClasspathResourceRepository.newBuilder()
-				.includeLocationPattern( "classpath*:**/profiles/**/*.xml" )
-				.build()
-				.findAll()
-				.flatMap( resource ->
-				{
-					var uri = resource.getUri();
-					try
-					{
-						var profile = factory.newProfile( uri );
-						return Stream.of( profile );
-					}
-					catch ( NotDocumentException | IOException e )
-					{
-						log.error( "Couldn't load profile from {}", uri, e );
-						return Stream.empty();
-					}
-				} )
-				.toList();
+
+		var resources = applicationContext.getResources( "classpath*:**/profiles/**/*.xml" );
+		var profiles = new ArrayList<Profile>(resources.length);
+		for ( var resource : resources )
+		{
+			log.debug( "Loading profile from \"{}\"", resource );
+			try( var inputStream = resource.getInputStream() )
+			{
+				var profile = factory.newProfile( inputStream );
+				profiles.add( profile );
+			}
+			catch ( NotDocumentException | IOException e )
+			{
+				log.error( "Couldn't load profile from \"{}\"", resource, e );
+			}
+		}
+
+		return profiles;
 	}
 
 	@Bean
@@ -121,31 +134,53 @@ public class Server extends SpringBootServletInitializer
 	}
 
 	@Bean
-	public List<Resource.V10> demoDocuments()
+	public List<Resource.V10> demoDocuments() throws IOException
 	{
 		log.info( "Discovering built-in documents" );
-        try
+		var resources = applicationContext.getResources("classpath*:**/demo-documents/ddi-v25/*.xml");
+		var excludedResources = applicationContext.getResources( "classpath*:**/demo-documents/ddi-v25/*profile*.xml");
+
+		var includedResourcesSet = new HashSet<>( Arrays.asList( resources ) );
+		for (var excluded : excludedResources)
 		{
-			return ClasspathResourceRepository.newBuilder()
-					.includeLocationPattern( "classpath*:**/demo-documents/ddi-v25/*.xml" )
-					.excludeLocationPattern( "classpath*:**/demo-documents/ddi-v25/*profile*.xml" )
-					.build()
-					.findAll()
-					.map( Resource::getUri )
-					.map( uri ->
-					{
-						var path = uri.toString();
-						var lastPathIndex = path.lastIndexOf( '/' );
-						var fileName = path.substring( lastPathIndex + 1 );
-						return newResource( uri, fileName );
-					} )
-					.map( Resource.V10.class::cast )
-					.toList();
+			// Remove profiles from the document list
+			includedResourcesSet.remove( excluded );
+			log.debug( "\"{}\" excluded", excluded );
 		}
-		catch ( RuntimeException e )
+
+		var resourcesWithFileNames = new ArrayList<Resource.V10>();
+
+		for ( var resource : includedResourcesSet )
 		{
-			log.warn( "Couldn't discover demo documents", e );
-			return Collections.emptyList();
+			log.debug( "Discovering document at \"{}\"", resource );
+			try
+			{
+				URI uri;
+				if ( resource instanceof UrlResource urlResource )
+				{
+					// Clean URL paths, needed on Windows as URIs don't accept '\'
+					var urlString = urlResource.getURL().toString();
+					var cleanedURL = StringUtils.cleanPath( urlString );
+					uri = new URI( cleanedURL );
+				}
+				else
+				{
+					uri = resource.getURI();
+				}
+
+				var fileName = resource.getFilename();
+
+				// This shouldn't be null if a URI is set
+				assert fileName != null;
+
+				resourcesWithFileNames.add( newResource( uri, fileName ) );
+			}
+			catch ( IOException | URISyntaxException e )
+			{
+				log.error( "Resource \"{}\" cannot be converted into a URI", resource, e );
+			}
 		}
+
+		return resourcesWithFileNames;
 	}
 }

--- a/src/main/java/eu/cessda/cmv/server/ui/ResourceSelectionComponent.java
+++ b/src/main/java/eu/cessda/cmv/server/ui/ResourceSelectionComponent.java
@@ -83,9 +83,27 @@ public class ResourceSelectionComponent<T> extends CustomComponent
 	}
 
 	public ResourceSelectionComponent(
+		SelectionMode selectionMode,
+		ProvisioningOptions selectedProvisioningOption,
+		List<T> predefinedResources,
+		ItemCaptionGenerator<T> stringMapper,
+		Function<Resource.V10, Optional<T>> resourceValidator)
+	{
+		this(
+			selectionMode,
+			selectedProvisioningOption,
+			predefinedResources,
+			stringMapper,
+			item -> new Label(stringMapper.apply( item )),
+			resourceValidator
+		);
+	}
+
+	public ResourceSelectionComponent(
             SelectionMode selectionMode,
             ProvisioningOptions selectedProvisioningOption,
             List<T> predefinedResources,
+			ItemCaptionGenerator<T> stringMapper,
 			Function<T, Label> labelMapper,
 			Function<Resource.V10, Optional<T>> resourceValidator)
 	{
@@ -101,7 +119,7 @@ public class ResourceSelectionComponent<T> extends CustomComponent
 
 		this.predefinedSelect = new NativeSelect<>();
 		this.predefinedSelect.setEmptySelectionCaption( bundle.getString( "comboBox.select" ) );
-		this.predefinedSelect.setItemCaptionGenerator( Object::toString );
+		this.predefinedSelect.setItemCaptionGenerator( stringMapper );
 		this.predefinedSelect.setWidth( 100, Unit.PERCENTAGE );
 		this.predefinedSelect.setItems( predefinedResources );
 		this.predefinedSelect.addSelectionListener( listener -> listener.getSelectedItem().ifPresent( this::selectResource ) );

--- a/src/main/java/eu/cessda/cmv/server/ui/ValidationView.java
+++ b/src/main/java/eu/cessda/cmv/server/ui/ValidationView.java
@@ -108,7 +108,7 @@ public class ValidationView extends VerticalLayout implements View
                 SINGLE,
                 BY_PREDEFINED,
                 demoProfiles,
-                profile -> new Label( profile.getProfileName() + ": " + profile.getProfileVersion() ),
+				profile -> profile.getProfileName() + ": " + profile.getProfileVersion(),
                 this::parseProfile
 		);
 		this.profileSelectionComponent.setCaption( bundle.getString( "configuration.profileSelectionCaption" ) );
@@ -118,6 +118,7 @@ public class ValidationView extends VerticalLayout implements View
                 MULTI,
                 BY_UPLOAD,
                 demoDocuments,
+				Resource.V10::getLabel,
                 ValidationView::labelFromResource,
                 this::recognizeDdiDocument
 		);

